flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointContext.java

```diff
@@ -0,0 +1,197 @@
+        return executionStates.stream()
+                .filter(state -> state != ExecutionState.FINISHED)
+                .collect(Collectors.toSet());
+    }
+
+    /**
+     * {@code StopWithSavepointState} represents the different states during the stop-with-savepoint
+     * operation.
+     *
+     * <p>The state transitions are implemented in the following way: InitialWait ->
+     * [WaitForSavepointCreation|WaitForJobTermination] -> Final
+     */
+    private enum StopWithSavepointState {
+        InitialWait {
+            @Override
+            protected StopWithSavepointState handleSavepointCreation(
+                    StopWithSavepointContext context, String path) {
+                context.path = path;
+                return WaitForJobTermination;
+            }
+
+            @Override
+            public StopWithSavepointState onExecutionsTermination(
+                    StopWithSavepointContext context, Collection<ExecutionState> executionStates) {
+                context.unfinishedStates = extractUnfinishedStates(executionStates);
+                return WaitForSavepointCreation;
+            }
+        },
+        WaitForSavepointCreation {
+            @Override
+            public StopWithSavepointState handleSavepointCreation(
+                    StopWithSavepointContext context, String path) {
+                Preconditions.checkState(
+                        context.unfinishedStates != null,
+                        InitialWait + " should have preceded: No unfinishedStates is set.");
+
+                if (!context.unfinishedStates.isEmpty()) {
+                    return context.terminateExceptionWithGlobalFailover(context.unfinishedStates);
+                }
+
+                return context.terminateSuccessfully(path);
+            }
+        },
+        WaitForJobTermination {
+            @Override
+            public StopWithSavepointState onExecutionsTermination(
+                    StopWithSavepointContext context, Collection<ExecutionState> executionStates) {
+                Preconditions.checkState(
+                        context.path != null,
+                        InitialWait + " should have preceded: No path is set.");
+
+                Collection<ExecutionState> unfinishedExecutionStates =
+                        extractUnfinishedStates(executionStates);
+                if (!unfinishedExecutionStates.isEmpty()) {
+                    return context.terminateExceptionWithGlobalFailover(unfinishedExecutionStates);
+                }
+
+                return context.terminateSuccessfully(context.path);
+            }
+        },
+        Final;
+
+        public StopWithSavepointState onSavepointCreation(
+                StopWithSavepointContext context, String path, Throwable throwable) {
+            if (throwable != null) {
+                return context.terminateExceptionally(throwable);
+            }
+
+            return handleSavepointCreation(context, path);
+        }
```

Comment:
I think this logic is not correct. If we reach `WaitForSavepointCreation` first, then we want to restart the job globally if there was a savepoint failure or any of the `Executions` is not in `FINISHED`.

Reply:
Alternatively, we should enforce that we must stick to the order `InitialWait -> WaitForJobCompletion -> Final`

Reply:
Unfortunately, we cannot expect the savepoint creation to complete before the `ExecutionStates` switch to a terminal state. Checkpoint abortion is triggered [after the resources of `Executions` are released](https://github.com/XComp/flink/blob/fe413286b8082c4089a05604da0aba285950563e/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java#L1181). Hence, we cannot rely on the Savepoint creation future terminating before the `Executions` terminate.

Reply:
But if we are sure that both signals will eventually happen, then we could order the signals before they are given to the `StopWithSavepointOperation`, right? I think by making the callbacks dependent on each other it could work.

Reply:
`StopWithSavepointManager` handles the order of completion now.